File: Cocoa/Classes/AttributesFormatter.m
#import "AttributesFormatter.h"
#import <Carbon/Carbon.h>

@implementation AttributesFormatter

- (NSString *)stringForObjectValue:(id)obj
{
	BOOL addComma = NO;
	short attributes = [obj shortValue];
	NSMutableString *string = [NSMutableString string];
	
	// there's probably a smarter, bitwise, way to do this
	short attributeCount = 0;
	if( attributes & resPreload )	attributeCount++;
	if( attributes & resProtected )	attributeCount++;
	if( attributes & resLocked )	attributeCount++;
	if( attributes & resPurgeable )	attributeCount++;
	if( attributes & resSysHeap )	attributeCount++;
	
	// create string
	if( attributes & resPreload )
	{
		if( addComma )				[string appendString:@", "];
		if( attributeCount > 2 )	[string appendString:@"Pre"];
		else						[string appendString:@"Preload"];
		addComma = YES;
	}
	if( attributes & resProtected )
	{
		if( addComma )				[string appendString:@", "];
		if( attributeCount > 2 )	[string appendString:@"Pro"];
		else						[string appendString:@"Protected"];
		addComma = YES;
	}
	if( attributes & resLocked )
	{
		if( addComma )				[string appendString:@", "];
		if( attributeCount > 2 )	[string appendString:@"L"];
		else						[string appendString:@"Locked"];
		addComma = YES;
	}
	if( attributes & resPurgeable )
	{
		if( addComma )				[string appendString:@", "];
		if( attributeCount > 2 )	[string appendString:@"Pur"];
		else						[string appendString:@"Purgeable"];
		addComma = YES;
	}
	if( attributes & resSysHeap )
	{
		if( addComma )				[string appendString:@", "];
		if( attributeCount > 2 )	[string appendString:@"Sys"];
		else						[string appendString:@"SysHeap"];
		addComma = YES;
	}
	return string;
}

- (NSAttributedString *)attributedStringForObjectValue:(id)obj withDefaultAttributes:(NSDictionary *)attrs
{
	NSString *string = [self stringForObjectValue:obj];
	return [[NSAttributedString alloc] initWithString:string attributes:attrs];
}

- (NSString *)editingStringForObjectValue:(id)obj
{
	return nil;
}

- (BOOL)getObjectValue:(id *)obj forString:(NSString *)string errorDescription:(NSString **)error
{
	return NO;
}

- (BOOL)isPartialStringValid:(NSString *)partialString newEditingString:(NSString **)newString errorDescription:(NSString **)error
{
	return NO;
}

@end

File: Classes/AttributesFormatter.swift
import Foundation
import RKSupport

class AttributesFormatter: Formatter {
    static let names: [Int: (full: String, short: String)] = [
        ResAttributes.preload.rawValue: ("Preload", "Pre"),
        ResAttributes.protected.rawValue: ("Protected", "Pro"),
        ResAttributes.locked.rawValue: ("Locked", "L"),
        ResAttributes.purgeable.rawValue: ("Purgeable", "Pur"),
        ResAttributes.sysHeap.rawValue: ("SysHeap", "Sys")
    ]
    
    override func string(for obj: Any?) -> String? {
        guard let attributes = obj as? ResAttributes else {
            return nil
        }
        let list = [.preload, .protected, .locked, .purgeable, .sysHeap].filter { attributes.contains($0) }
        return list.map {
            let name = Self.names[$0.rawValue]!
            return list.count > 2 ? name.short : name.full
        }.joined(separator: ", ")
    }
}
